src/core/include/lattice/hal/default/dcrtpoly-impl.h

```diff
@@ -808,173 +778,83 @@ void DCRTPolyImpl<VecType>::ModReduce(const NativeInteger& t, const std::vector<
-    // Precompute the Barrett mu parameter
-    Integer mu = bigModulus.ComputeMu();
-
-    // now, compute the values for the vector
-    for (usint ri = 0; ri < ringDimension; ri++) {
-        coefficients[ri] = 0;
-        for (usint vi = 0; vi < nTowers; vi++) {
-            coefficients[ri] += (Integer((*vecs)[vi].GetValues()[ri].ConvertToInt()) * multiplier[vi]);
-        }
-        coefficients[ri].ModEq(bigModulus, mu);
+#pragma omp parallel for private(tmp1)
+    for (uint32_t j = 0; j < r; ++j) {
+        for (uint32_t i = 0; i < t; ++i)
+            V[j] += (tmp1 = m_vectors[i].GetValues()[j].ConvertToInt()) * multiplier[i];
+        V[j].ModEq(qt);
     }
 
     // Setting the root of unity to ONE as the calculation is expensive and not required.
-    DCRTPolyImpl<VecType>::PolyLargeType polynomialReconstructed(
-        std::make_shared<ILParamsImpl<Integer>>(m_params->GetCyclotomicOrder(), bigModulus, 1));
-    polynomialReconstructed.SetValues(std::move(coefficients), Format::COEFFICIENT);
-
-    return polynomialReconstructed;
+    DCRTPolyImpl<VecType>::PolyLargeType poly(std::make_shared<ILParamsImpl<Integer>>(2 * r, qt, 1));
+    poly.SetValues(std::move(V), Format::COEFFICIENT);
+    return poly;
 }
 
 /*
- * This method applies the Chinese Remainder Interpolation on a
- * single element across all towers of a DCRTPolyImpl and produces an Poly
- * with zeros except at that single element
- * How the Algorithm works:
- * Consider the DCRTPolyImpl as a 2-dimensional matrix M, with dimension
- * ringDimension * Number of Towers. For brevity , lets say this is r * t Let
- * qt denote the bigModulus (all the towers' moduli multiplied together) and
- * qi denote the modulus of a particular tower. Let V be a BigVector of size
- * tower (tower size). Each coefficient of V is calculated as follows: for
- * every r calculate: V[j]= {Sigma(i = 0 --> t-1) ValueOf M(r,i) * qt/qi *[
- * (qt/qi)^(-1) mod qi ]}mod qt
- *
- * Once we have the V values, we construct an Poly from V, use qt as it's
- * modulus, and calculate a root of unity for parameter selection of the Poly.
+ * This method applies Chinese Remainder Interpolation on a
+ * single element across all towers of a DCRTPolyImpl and produces
+ * an Poly with zeros except at that single element
  */
 template <typename VecType>
 typename DCRTPolyImpl<VecType>::PolyLargeType DCRTPolyImpl<VecType>::CRTInterpolateIndex(usint i) const {
-    usint ringDimension = m_params->GetRingDimension();
-    usint nTowers       = m_vectors.size();
-    Integer bigModulus(m_params->GetModulus());  // qT
-
-    VecType coefficients(ringDimension, bigModulus);
-
-    std::vector<Integer> multiplier(nTowers);
-    for (usint vi = 0; vi < nTowers; vi++) {
-        Integer qj(m_vectors[vi].GetModulus().ConvertToInt());
-        Integer divBy  = bigModulus / qj;
-        Integer modInv = divBy.ModInverse(qj).Mod(qj);
-        multiplier[vi] = divBy * modInv;
-    }
+    if (m_format == Format::EVALUATION)
```

Comment:
could it be replaced with "if (m_format != Format::COEFFICIENT)"?

Reply:
will update in PR for issue 458